vlasovsolver/cpu_trans_pencils.cpp

```diff
@@ -0,0 +1,1170 @@
+               nextNeighbor = selectNeighbor(grid,id,dimension,newPath);
+
+               if ( newPath == 3 ) {
+                  // This builder continues with neighbor 3 with an extended path
+                  path = myPath;
+                  if(nextNeighbor != INVALID_CELLID) {
+                     coordinates = grid.get_center(nextNeighbor);
+                  }
+               } else {
+                  // Spawn recursive new builders for neighbors 0,1,2
+                  buildPencilsWithNeighbors(grid,pencils,id,ids,dimension,myPath,endIds);
+               }
+            }
+         }
+      } // Closes if (refLvl == 0)
+
+      // If we found a neighbor, let's verify if it should be translated
+      if(nextNeighbor != INVALID_CELLID) {
+         if (debug) {
+            std::cout << " Next neighbor is " << nextNeighbor << "." << std::endl;
+         }
+         // Non-local, non-translated, and ids belonging to other pencils are not included
+         if ( std::any_of(endIds.begin(), endIds.end(), [nextNeighbor](uint i){return i == nextNeighbor;}) ||
+              !do_translate_cell(grid[nextNeighbor])) {
+            nextNeighbor = INVALID_CELLID;
+         } else {
+            // Yep, this goes in this pencil.
+            ids.push_back(nextNeighbor);
+         }
+      }
+
+      id = nextNeighbor;
+   } // Closes while loop - end of pencil reached.
+
+   // Get the x,y - coordinates of the pencil (in the direction perpendicular to the pencil)
+   double x,y;
+   int ix=0, iy=0;
+
+   switch(dimension) {
+   case 0:
+      ix = 1;
+      iy = 2;
+      break;
+   case 1:
+      ix = 0;
+      iy = 2;
+      break;
+   case 2:
+      ix = 0;
+      iy = 1;
+      break;
+   }
+   //ix = (dimension + 1) % 3; // incorrect for DCCRG
+   //iy = (dimension + 2) % 3;
+
+   x = coordinates[ix];
+   y = coordinates[iy];
+
+   pencils.addPencil(ids,x,y,periodic,path);
+   return;
+}
+
+/* Determine which cells in the local DCCRG mesh should be starting points for pencils.
+ * If a neighbor cell is non-local, across a periodic boundary, or in non-periodic boundary layer 1
+ * then we use this cell as a seed for pencils
+ *
+ * @param [in] mpiGrid DCCRG grid object
+ * @param [in] localPropagatedCells List of local cells that get propagated
+ * ie. not L2-boundary or DO_NOT_COMPUTE
+ * @param [in] dimension Spatial dimension
+ * @param [out] seedIds list of cell ids that will be starting points for pencils
+ */
+void getSeedIds(const dccrg::Dccrg<SpatialCell,dccrg::Cartesian_Geometry>& mpiGrid,
+                const vector<CellID> &localPropagatedCells,
+                const uint dimension,
+                vector<CellID> &seedIds) {
+
+   const bool debug = false;
+   int myRank;
+   if (debug) MPI_Comm_rank(MPI_COMM_WORLD,&myRank);
+
+   // These neighborhoods now include the AMR addition beyond the regular vlasov stencil
+   int neighborhood = getNeighborhood(dimension,VLASOV_STENCIL_WIDTH);
+
+#pragma omp parallel for
```

Comment:
Same question, scheduling?

Reply:
This is again a lightweight loop but with non-constant computational cost, so I think regular dynamic schedule is fine.